### ISSUE TITLE
PHP: better debug info for Message objects in C-extension

### ISSUE THREAD

**Issue by bshaffer** [protocolbuffers/protobuf#14872]
See https://github.com/protocolbuffers/protobuf/issues/12714 and https://github.com/protocolbuffers/protobuf/pull/12718, which resolved this problem for the native library.

When the protobuf c-extension is enabled for PHP, calling `var_dump` on a protobuf message object does not output any useful information:

```php
// Calling var_dump on a Protobuf message (extension enabled):
php > $timestamp = new Google\Protobuf\Timestamp();
php > $timestamp->setSeconds(12345);
php > var_dump($timestamp);
object(Google\Protobuf\Timestamp)#1 (0) {
}
```

Ideally this would output something like this:

```
object(Google\Protobuf\Timestamp)#12 (1) {
  ["seconds"]=>
  int(12345)
}
```

**Comment by northmule** [comment]
A similar problem. Checked on 3.12.4, 3.17, 3.24.4
php.ini
```
extension=protobuf.so
```

**Comment by bshaffer** [comment]
@northmule what is the problem you're seeing? 

For clarification, this issue is a feature request for the extension to produce better debugging info. It has not yet been implemented.

**Comment by northmule** [comment]
The problem is that when connecting From the library in the php.ini file, the ability to view the class object inherited from the parent class \Google\Protobuf\Internal\Message disappears.
I have prepared two visual screenshots from the Phpstorm IDE and the configured xDebug debugger.
In one screenshot in php.ini, the library connection is disabled - ;extension=protobuf.so
In another screenshot, the library connection is enabled - extension=protobuf.so
```
extension=protobuf.so
```
![Screenshot_20231212_084650](https://github.com/protocolbuffers/protobuf/assets/33060300/cba1db80-16cb-4bb6-a8d9-56546b54f8db)


```
;extension=protobuf.so
```
![Screenshot_20231212_084512](https://github.com/protocolbuffers/protobuf/assets/33060300/53a7a7c0-0fe4-4e78-8be3-3afcfe430316)

I highlighted the object in red.

If the library is not connected, I can view the object, it is filled with data

If the library is connected, I cannot view the object.

**Comment by bshaffer** [comment]
I assume this is because in the first example, the object is implemented in the C extension, and so is not visible by the IDE. This feature request is to implement a hook in the message class implemented in C++ which, similar to `__debugInfo` in PHP, when `var_dump` is called, it prints the properties inside the C object.

**Comment by haberman** [comment]
I think this is a reasonable feature request, and we'd be happy to accept a contribution.

This would involve implementing [`get_debug_info`](https://wiki.php.net/internals/engine/objects#get_debug_info) for `Message`, `RepeatedField`, and `MapField`.

There was previously a PR to return an empty hash table for `get_debug_info`, which is a good example of how this might look, except that we'd want the returned hash table to be fully populated: https://github.com/protocolbuffers/protobuf/pull/7562/files

**Comment by haberman** [comment]
I just noticed that we have a test for debug info that is disabled for the C extension: https://github.com/protocolbuffers/protobuf/blob/6f1d88107f268b8ebdad6690d116e74c403e366e/php/tests/DebugInfoTest.php#L37-L42

So we essentially already have a spec for how the C extension should work in this regard.

**Comment by bshaffer** [comment]
@haberman This isn't a justification for the existing behavior, it's just confirming that it isn't currently supported.

**Comment by haberman** [comment]
> I think you're just saying that the C-extension's implementation can follow the "spec" of the native implementation (and pass the tests), and I totally agree.

Exactly. I meant it as a hint/pointer for whoever ends up implementing this for C. The easiest way to start is to re-enable the test, and code a solution until the test passes.
